### DISCUSSION TITLE
[Draft] Feature Request: Further developments TEKSI Wastewater & GEP

### DISCUSSION THREAD

**Discussion by sjib** [teksi/wastewater#240] [Ideas]
## General information

In 2020 the Verband Schweizer Abwasser- und Gewässerschutzfachleute (VSA) published a new release of its datamodels (SIA405 Abwasser, VSA-DSS and VSA-KEK). Release 2020 replaces release 2015 (including the extensions ALR, VSA-KEK, Massnahme, Melioration, Stammkarte). Release 2020 reduces the numbers of classes and themes (such as rivers, accidents, hazardous substances that are covered by other datamodels (from cantons or the government) and includes the extensions (ALR, Massnahme, Melioration and Stammkarte) into the main model (VSA-DSS). VSA-KEK is still published as an extension of SIA405 Abwasser. 

TEKSI wastewater and GEP Release 2024.0 builds on VSA-DSS 2020.1 Release. But in that release not all new classes are supported yet and can be easily used for GEP mandates

- Applicant:  **please add your organisation here if you are interested in this**
- Developers:  to be defined 
- User group referent: ?
- Developments referent: ?
- Project Manager: to be defined

## Users and administrators needs
- TEKSI wastewater and GEP users should be able to work with the release 2020.1 of the VSA data models as it brings in new themes that are important in the GEP management. Cantons and municipalities have started to require release 2020 for new projects and GEP management.

### Description of users and administrators feature requirements
- The new themes of VSA-DSS 2020 Measure (Massnahme), Log File (Stammkarte) and ALR (Abwasserentsorgung im ländlichen Raum) should be supported with (basic) user forms. 
- The current functionalities were integrated in the first phase with one exception: **length profile**. This should be re-integrated and be rebuild on a new technological basis, as the qgep code was not easy transferable.

#### Specific requirements
The project is quite big therefore it is suggested to **split in different steps phases**:

List of features:

- [ ] Complete French translation (project file)
- [ ] Complete documentation English, French, German with new features especially creation of links with weirs etc.
- [ ] Extra colors scheme as used by the VSA Wiki (including special structures).
- [ ] Redevelop length profile tool
- [x] [Setting default dataowner and provider](https://github.com/teksi/wastewater/pull/173)
- [ ] Language dependencies handling
- [x] Datamodel tool for installation and upgrades -> [separate project for maintaining all TEKSI modules](https://github.com/teksi/TMMT)
- [ ] Add subsystem Measure (Massnahme) including relations
- [x] Add subsystem Log_file (Stammkarten)
- [x] Add subsystem ALR (Abwasserentsorgung im ländlichen Raum)
- [ ] ... 

##### New feature: fields dataowner and provider
- [x] There should be a concept how to handle these attributes automatically by providing default values. 
- [x] Proposal: define default-values for new records in the qgep_sys-schema, e.g. as new fields in existing or new sys table or in TEKSI project configuration (project variable?)
- [x] Implementation see https://github.com/teksi/wastewater/pull/173 (already merged and documented)

##### Language dependencies and handling
- [ ] Currently a _de.qm file is considered to display the project file language - with the new qgz format it is not so clear what really happens 
- [ ] Check handling of language files with new qgz format 
- [ ] Are there advantages of qgz format that we need - else set default back to qgs
- [ ]  Check whether project language should be stored in sys table also?
- [ ] update documentation about this
- [x] check all code on language dependencies - e.g. layer names vw_qgep_reach etc. and adapt code to be language independant - examples see https://github.com/QGEP/qgepqwat2ili/issues/92

#### Add subsystem Measure (Massnahme) including relations to other classes
New theme Massnahme and attributes will need to be configured if they need special widget configurations
![massnahme](https://user-images.githubusercontent.com/1265865/211002305-7527f61c-9cbe-4e1a-9807-afed58be2011.png)

- can be connected via maintenance_event to wastewater_structures (n:m relation possible)
- and can have the geometry of these wastewater_structures) 
- or can be connected to Gebäudegruppe 
- or can stand alone and a geometry can be defined free. 

- [ ] Define use cases
- [ ] Define geometry similar to wastewater_structure with action? 
- [ ] UI in plugin!

**Delivarables**: 
- [ ] New project file with extended functionality (attribute forms) and extent
- [ ] Updated user documentation for new classes, forms and features
- [ ] Updated Demodata set with Measures related to wastewater_structures using n-m relation of maintenance_event

#### Add subsystem Log_file (Stammkarten)
New theme log_card (Stammkarte) and attributes will need to be configured if they need special widget configurations

![subsystem_stammkarte](https://user-images.githubusercontent.com/1265865/211002538-45270337-9eb5-427b-b6b7-5cc23fb2ca4b.png)

###### new class Biol_oeko_Gesamtbeurteilung

- [ ] need of new UI for maintenance_event and subclasses examination, Unterhalt and bio-oek-Gesamtbeurteilung -> vw_examination, vw_bio_oek_gesamtbeurteilung, vw_maintenace instead of vw_qgep_maintenance. New relation to class new Massnahme. 
- [ ] Compare two options of implementation: Either as one big view (difficult to overlook) or as three subviews (prioritized option)

###### new class Stammkarte
- [x] create vw_log_file for this class
- [ ] further UI to be developed later

**Delivarables**: 
- New project file with extended functionality (attribute forms) and extent
- Updated user documentation for new classes, forms and features
- Updated Demodata set with log_files for all log_file types and Bauwerkskomponenten

#### Add subsystem ALR (Abwasserentsorgung im ländlichen Raum) 
New theme ALR and attributes will need to be configured if they need special widget configurations

![subsystem_alr](https://user-images.githubusercontent.com/1265865/211002615-071b3ee7-1b7f-4384-b06f-4beb2dc1b3a5.png)

###### new classes ALR

- [ ] KLARA (creates also wastewater node, but not cover, separate layer, not part of vw_qgep_wastewater_structure) 
- [ ] Abflusslose_Toilette (with wastewater node needed? will not have network connection)
- [ ] Entsorgung, Gebäudegruppe, Landwirtschaftsbetrieb, Gebäudegruppe_BAUGWR) 
   ** [ ] No special implementation in plugin or project.
   ** [ ] create views for these classes
 - [ ] further UI to be developed later

####  New class reach_alternative_progression (Haltung_Alternativverlauf)
_(visible in Klassendiagramm and ili, not visible in Objektkatalog)._ 

![haltung_alternativverlauf](https://user-images.githubusercontent.com/1265865/211002649-31ea8745-4331-41f1-8dbd-e984a13d58de.png)

- [ ] Implementation of class and relation in datamodel

![dss2020_haltung_alternativverlauf](https://user-images.githubusercontent.com/1265865/211012876-683275d6-f065-44c5-b912-61f4d4055cb4.png)

- [ ] ui to define the geometry (parallel to Verlauf?) with scale selection
- [ ] ui/configuration for the symbolisation - in function of scale?

###### Subsystem Messung
_not really implemented_

- [ ] If a user wants to use this also for calculation results in the paa-network etc., then there should be special UI -> define follow up project

![subsystem_messung](https://user-images.githubusercontent.com/1265865/211012984-295de7d1-f953-4a1f-ba7b-f2645d3022e1.png)

**Delivarables**: 

- [x] Updated datamodel with alternative_progression
- [ ] Adapted plugin with new functionalities
- [ ] New project file with extended functionality (attribute forms) and extent
- [ ] Updated user documentation for new classes, forms and features (Alternativ_Verlauf)
- [ ] Updated Demodata set with measures and alternative_progression (Alternativ_Verlauf) in different scales

### Illustrations of the features and the requirements

_Schemes, screen captures, or any other illustrations._

### Feasibility study

Feasibility, options, alternatives and technical specificities

### Risks and constraints
The VSA has published the Release 2020 together with the information, that Release 2015 will only be supported till end of 2023. Therefore it is a high priority that TEKSI adapts it's datamodel to release 2020 as fast as possible. If not TEKSI risks to build on a datamodel released that is not supported by the VSA anymore.

### Mutualisation 
_Mutualisation and possible impacts with other projects (QGIS, QWAT, QGEP)_

## Specification and offers ( Developer )
Not shared on Github.

### Tenders

Not shared on Github.

### Timeline / deadlines (draft)
_Add a detailed timeline_

- [x] 9.4.2024 : Start defining feature request so it complies for 3_professional_validation (fachlich) and 4_strategic_validation
- [ ] 5.2024 : Feature definition completed
- [ ] 5.2024 : Strategic validation by board
- [ ] 6.2024 : 5_Technical validation by technical group
- [ ] Requesting 6_Offer
- [ ] 7_Validation_of_offer
- [ ] 8_Development Phase 1
- [ ] 9_Control_of_development Phase 1
- [ ] 10_Integration Phase 1
- [ ] 11_Installation Phase 1
- [ ] 12_Completion Phase 1
- [ ] 8_Development Phase 2
- [ ] 9_Control_of_development Phase 2
- [ ] 10_Integration Phase 2
- [ ] 11_Installation Phase 2
- [ ] 12_Completion Phase 2
etc

- [ ] Warranty period : to be adressed

### Related softwares versions

- QGIS version: 3.34.4 or later
- Database model version: 2024.0
- INTERLIS Model version: 2020.1 (see https://www.vsa.ch/models/?dir=2020_1)

Public issues links : 
- …

Helpdesk issues links : 
- …

### Realisation dates

- Tender reception date : 
- Tender validation date : 

## Documentation

Documentation links
- User manual : 
- Administrator manual :

## Development and integration

Link and comments from the developer about the product

- Deliverable : 
- Recommendations and comments : 
- Future prospects : 

## Conclusion

Conclusions from applicant, users or admins group

- Conclusion: 
- Future prospects:

**Description :** 

**Useful links :**
www.vsa.ch/wiki -> Blaudrucke
Model repository: https://www.vsa.ch/models

**Comment by urskaufmann** [comment]
Measure (Massnahme) seems to be quit easy to config (just the table, just an addition in the existing maintenance_event-configuration and later in the ALR-part). But a measure can have point-, line- or perimeter-geometry. How can this be handled within QGIS? Must there be three layers? Is it possible to change the layer in QGIS (e.g. delete the line_geometry and add a perimeter-geometry)?

**Reply by sjib** [reply]
Hallo Stefan,

Du bist sicher der beste Ansprechpartner.
Hier die Liste der «musts» um die Version 2020 zu benutzen.
Gibt garantiert issues und features die existieren, aber ich weiss nicht wo und welche.

TEKSI, what I still need:

- [ ] **Data model tool**
- A need!
- Create new empty data model 2020
- Apply delta of data model (minor updates installation).
- Create new Qgis-file in French / german.

- [ ] **Interlis:**

- Chose export language (French / German) -> Could also be done in the GEP-Datachecker?? .

- [ ] **No more dependencies from python to install.**

- [ ] **Qgis file:**

o	New formulars for new classes and attributes.
o	New themes (ALR, Stammkarte)
o	Complete traduction in French / german.
o	Colors and names as used by the VSA Wiki (including special structures).
o	The size of the canalisations and manholes should be visible when you zoom.

- [ ] **Network following and profile tool working.**
- [ ] **Documentation in French**
- [ ] **Easily create a hydraulic link between to nodes** (for special structures) -> Feature:
o	Chose to create new link in the wizard.
o	Click on the first node then the second (which can be also a reach _point), check:
        - First one is a node.
        - Second one does not have the same coordinates.
        - Second one is either node or a reach_point. 
o	Create:
        - the secondary node:
               * Linked to the primary node.
               * Take the altitude from the reach point. 
        - The link (which is in the primary node) 
o	 and show it.
o	When you click the “I” in , the link description should be shown and be modified.

Wenn diese Sachen funktionieren, wird TEKSI gebraucht werden.

Ich habe auf dem VSA-Seminar (https://vsa.ch/wp-content/uploads/2023/11/30-31.01.2024-VSA_Saisie_2024_flyer_v4-NB01-VSA-1123.pdfgarantiert ) einige von TEKSI überzeugt (mit meiner Live-präsentation von TEKSI für Sonderbauwerke).

Moritz

**Reply by sjib** [reply]
> Chose export language (French / German) -> Could also be done in the GEP-Datachecker?? .

Anleitung für die Übersetzung hier: https://www.sjib.ch/wie-uebersetze-ich-eine-interlis-transferdatei-in-eine-andere-sprache/
Für VSA-DSS 2020 braucht es allenfalls auch Baskets.
Diese Konzept könnte man gut in die Schnittstelle einbauen - es basiert auch auf ili2db

**Reply by cymed** [reply]
In #187 I introduced three new layers, which works just fine. If you want to add a geometry to an existing entry, select it in the attribute table and use the "add Part" tool

**Comment by urskaufmann** [comment]
Wishes for profile tool:
- the old tool draw not always the data in the database (e.g. if there was level = NULL oder level = 0). Lock at QGEP Issue#464.  If there is a missing cover.level or a missing node.bottom_level and missing reachpoint.levels, then there is no line.
- in the old tool, there was obj_id and not identifier shown in the box for reaches or manholes. The fields to be shown must be defined.
- for online-use was the text good. For a printversion of the length-profile, there should be the text for manholes and reaches (identifier, levels, material, slope, length, clear_height) on the bottom of the picture

**Comment by urskaufmann** [comment]
Selections should be optimized: 
if I use network-following today, I can choose "select current path". This selects the found records in the layers vw_tww_reach, vw_wastewater_node and vw_tww_wastewater_structure. 
1. The command "select current path" has nothing to do with the length profile. There should be a better place to start this function.
2. When changing a selection in QGIS eg. by formula, I can choose, whether there should be a new selection, add to an existing selection, remove from existing selection or select in the existing selection. "Select current path" should have the same options (a must!)
3. The command should also select the connected records (connected to the selected networkelements) in the layer catchment_area and vw_individual_surface. This functionality can be optional, because in big datasets it can probably take some time. I think this layers are the most important.
4. I would call this command "select connected records" and separate it totaly from the actual "select current path". The function is: Select connected records starts always with the current selection in layer vw_tww_reach.
5. Like this, "select current path" must only select records in vw_tww_reach, no more in other layers. The only reason to not automatically start this selection with the networkfollowing are the select-options (new selection,add selection,remove from selection,select in selection) ?
6. Generally the same records are selected as when exporting a selection. Because the code for this export-selections will be soon ready (I really hope so), it can/should be used also for this "select connected records".
7. As a benefit, we can start the "select connected records" before an export an see then before the export, which nodes, structures, catchment_areas etc will be exported (we called this wysiwyg years ago ;-) ).
conclusion: new select options for network following. new function select connected records, also with catchment area, individual surface

**Reply by sjib** [reply]
There code for this is in the qgepplugin qgepprofiledockwidget.py
https://github.com/QGEP/qgepplugin/blob/master/qgepplugin/gui/qgepprofiledockwidget.py#L159 (Select current path Button of the Profile toll with `onSelectCurrentPathAction`) and 

I guess from the swmm Export `onConfigureSelectAction`
https://github.com/QGEP/qgepplugin/blob/b122d593e99b1f4accb522db4f2c00758f547db0/qgepplugin/gui/qgepprofiledockwidget.py#L110

We would have to look how to adapt those functions towards what you need and maybe place this also to seperate extra buttons.

**Comment by sjib** [comment]
UI for wws.main_cover - see https://github.com/orgs/teksi/discussions/110

**Comment by sjib** [comment]
**LogCards are quiet a complex thing:**
There is a great potential for development: create log-cards automatic depending on the function of special_structures, connect the catchment_areas with the log_cards using the network following automatically, create hydraulic_char_data with status current automatic for logcards who need this record and create the catchment_area_totals – records automatical for this hydraulic_char_data-records.
Meanwile the manual workflow has to be well documented…
Please correct me, if I am on a wrong way…

**Hint:** 
there is an error in the DSS. With the actual modell you can not create more than one hydraulic_char_data that has the same node as fk_primary_direction. Means specially, you can not create different status on the same node… 

```
!!@ comment = "Assoziation"
ASSOCIATION Hydr_Kennwerte_PrimaerrichtungAssoc =
  PrimaerrichtungRef  -- {0..1} Abwasserknoten;
!!@comment = " nur für Trennbauwerke"
  Hydr_Kennwerte_PrimaerrichtungAssocRef -- {0..1} Hydr_Kennwerte;
END Hydr_Kennwerte_PrimaerrichtungAssoc;
```

_Das INTERLIS sagt, dass PrimaerrichtungRef optional ist – das ist ok
_Und dass nur maximal ein Element pro Abwasserknoten darauf verweisen darf – das müsste aber, wie z.B. bei den Einzugsgebieten 0..* sein._

_Hydr_Kennwerte_PrimaerrichtungAssocRef should associate 0 to 1 target objects (instead of 2) – instead of 2 sagt, dass du 2 mal den Verweis auf den gleichen Abwasserknoten hast, aber es müsste gemäss aktuellem Modell nur 0 oder 1 Verweise sein: {0..1}._

_So wäre es wohl korrekt:_
`  Hydr_Kennwerte_PrimaerrichtungAssocRef -- {0..*} Hydr_Kennwerte;`


Freundliche Grüsse
Urs Kaufmann

Originally posted https://github.com/teksi/wastewater/pull/712#issuecomment-3178426914

**Reply by sjib** [reply]
First idea of visualisation of log cards is available here: https://github.com/teksi/wastewater/pull/732#issuecomment-3234611597

**Reply by cymed** [reply]
Please also note that we would need a switch to disable the creation of log cards - they might be managed by another entity on another DB
